src/libsystemd-network/ndisc-option.c

```diff
@@ -731,6 +743,136 @@ static int ndisc_option_parse_prefix64(Set **options, size_t offset, size_t len,
         return ndisc_option_add_prefix64(options, offset, prefixlen, &prefix, lifetime);
 }
 
+int ndisc_option_add_encrypted_dns(Set **options, size_t offset, sd_dns_resolver *res, usec_t lifetime) {
+        assert(options);
+
+        sd_ndisc_option *p = ndisc_option_new(SD_NDISC_OPTION_ENCRYPTED_DNS, offset);
+        if (!p)
+                return -ENOMEM;
+
+        p->encrypted_dns = (sd_ndisc_dnr) {
+                .resolver = res,
+                .lifetime = lifetime,
+        };
+
+        return ndisc_option_consume(options, p);
+}
+
+static int ndisc_get_dns_name(const uint8_t *optval, size_t optlen, char **ret) {
+        _cleanup_free_ char *name = NULL;
+        int r;
+
+        assert(optval || optlen == 0);
+        assert(ret);
+
+        r = dns_name_from_wire_format(&optval, &optlen, &name);
+        if (r < 0)
+                return -EBADMSG; /* ndisc doesn't handle other errcodes atm */
+        if (r == 0 || optlen != 0)
+                return -EBADMSG;
+
+        *ret = TAKE_PTR(name);
+        return r;
+}
+
+static int ndisc_option_parse_encrypted_dns(Set **options, size_t offset, size_t len, const uint8_t *opt) {
+        int r;
+
+        assert(options);
+        assert(opt);
+        _cleanup_(sd_dns_resolver_done) sd_dns_resolver res = {};
+        usec_t lifetime;
+        size_t ilen;
+
+        /* Every field up to and including adn must be present */
+        if (len < 2*8 )
+                return -EBADMSG;
+
+        if (opt[0] != SD_NDISC_OPTION_ENCRYPTED_DNS)
+                return -EBADMSG;
+
+        size_t off = 2;
+
+        /* Priority */
+        res.priority = unaligned_read_be16(opt + off);
+        /* Alias mode is not allowed */
+        if (res.priority == 0)
+                return -EBADMSG;
+        off += sizeof(uint16_t);
+
+        /* Lifetime */
+        lifetime = unaligned_read_be32(opt + off);
+        off += sizeof(uint32_t);
+
+        /* adn field (length + dns-name) */
+        ilen = unaligned_read_be16(opt + off);
+        off += sizeof(uint16_t);
+        if (off + ilen > len)
+                return -EBADMSG;
+
+        r = ndisc_get_dns_name(opt + off, ilen, &res.auth_name);
+        if (r < 0)
+                return r;
+        if (dns_name_is_root(res.auth_name))
+                return -EBADMSG;
+        off += ilen;
+
+        if (off == len) /* adn-only mode */
+                return 0;
```

Comment:
If the ADN-only mode is used `off` can be less than `len` when options are padded with zeros. I sent
```sh
0000  90 03 00 01 00 00 00 FF 00 09 01 48 05 6C 6F 63  ...........H.loc
0010  61 6C 00 00 00 00 00 00                          al......
```
and it got past that `if` statement because `off` was 19 and `len` was `24`. `res` itself was correct
```
$2 = {priority = 1, auth_name = 0x61500000dc80 "H.local", family = 0,
addrs = 0x0, n_addrs = 0, transports = 0, port = 0, dohpath = 0x0}
```

(Also I think networkd should log something to make it clear that the ADN-only mode isn't supported)

Reply:
Ah, right.

A log message sounds reasonable too, I'll add one.